src/sync/ms_queue.rs

```diff
@@ -34,49 +68,209 @@ impl<T> MsQueue<T> {
+                    }
+                    Cache::Node(n) => n
+                }
+            }
+
+            /// Extract the data from the cache, deallocating any cached node.
+            fn into_data(self) -> T {
+                match self {
+                    Cache::Data(t) => t,
+                    Cache::Node(node) => {
+                        match node.into_inner().payload {
+                            Payload::Data(t) => t,
+                            _ => unreachable!(),
+                        }
+                    }
+                }
+            }
+        }
+
+        let mut cache = Cache::Data(t); // don't allocate up front
         let guard = epoch::pin();
+
         loop {
+            // We push onto the tail, so we'll start optimistically by looking
+            // there first.
             let tail = self.tail.load(Acquire, &guard).unwrap();
-            if let Some(next) = tail.next.load(Acquire, &guard) {
-                self.tail.cas_shared(Some(tail), Some(next), Release);
-                continue;
-            }
 
-            match tail.next.cas_and_ref(None, n, Release, &guard) {
-                Ok(shared) => {
-                    self.tail.cas_shared(Some(tail), Some(shared), Release);
-                    break;
+            // Is the queue in Data mode (empty queues can be viewed as either mode)?
+            if tail.is_data() ||
+                self.head.load(Relaxed, &guard).unwrap().as_raw() == tail.as_raw()
+            {
+                // Attempt to push onto the `tail` snapshot; fails if
+                // `tail.next` has changed, which will always be the case if the
+                // queue has transitioned to blocking mode.
+                match self.push_internal(&guard, tail, cache.into_node()) {
+                    Ok(_) => return,
+                    Err(n) => {
+                        // replace the cache, retry whole thing
+                        cache = Cache::Node(n)
+                    }
+                }
+            } else {
+                // Queue is in blocking mode. Attempt to unblock a thread.
+                let head = self.head.load(Acquire, &guard).unwrap();
+                // Get a handle on the first blocked node. Racy, so queue might
+                // be empty or in data mode by the time we see it.
+                let request = head.next.load(Acquire, &guard).and_then(|next| {
+                    match next.payload {
+                        Payload::Blocked(signal) => Some((next, signal)),
+                        Payload::Data(_) => None,
+                    }
+                });
+                if let Some((blocked_node, signal)) = request {
+                    // race to dequeue the node
+                    if self.head.cas_shared(Some(head), Some(blocked_node), Release) {
+                        unsafe {
+                            // signal the thread
+                            ptr::write(&mut (*signal).data, Some(cache.into_data()));
```

Comment:
With this using `Option<T>` now, this `ptr::write` could just be a normal assignment?